lua/lsp/null-ls.lua

```diff
@@ -1,141 +1,136 @@
+  if vim.fn.executable(command) == 1 then
+    return command
   end
-  table.insert(lvim.lang[ft][operation]._failed_requests, provider)
+  return nil
 end
 
-local function validate_nodejs_provider(provider)
-  local command_path
-  local root_dir
-  if lvim.builtin.rooter.active then
-    --- use vim-rooter to set root_dir
-    vim.cmd "let root_dir = FindRootDirectory()"
-    root_dir = vim.api.nvim_get_var "root_dir"
-  else
-    --- use LSP to set root_dir
-    local ts_client = require("utils").get_active_client_by_ft "typescript"
-    if ts_client == nil then
-      Log:get_default().error "Unable to determine root directory since tsserver didn't start correctly"
-      return
-    end
-    root_dir = ts_client.config.root_dir
+local function adapt_linter(linter)
+  -- special case: fallback to "eslint"
+  -- https://github.com/jose-elias-alvarez/null-ls.nvim/blob/9b8458bd1648e84169a7e8638091ba15c2f20fc0/doc/BUILTINS.md#eslint
+  if linter.exe == "eslint_d" then
+    return null_ls.builtins.diagnostics.eslint.with { command = "eslint_d" }
   end
-  local local_nodejs_command = root_dir .. "/node_modules/.bin/" .. provider._opts.command
-  Log:get_default().debug("checking for local node module: ", vim.inspect(provider))
-
-  if vim.fn.executable(local_nodejs_command) == 1 then
-    command_path = local_nodejs_command
-  elseif vim.fn.executable(provider._opts.command) == 1 then
-    Log:get_default().debug("checking in global path instead for node module", provider._opts.command)
-    command_path = provider._opts.command
-  else
-    Log:get_default().debug("Unable to find node module", provider._opts.command)
-  end
-  return command_path
+  return null_ls.builtins.diagnostics[linter.exe]
 end
 
-local function validate_provider_request(provider)
-  if provider == "" or provider == nil then
-    return
-  end
-  -- NOTE: we can't use provider.name because eslint_d uses eslint name
-  if vim.tbl_contains(nodejs_local_providers, provider._opts.command) then
-    return validate_nodejs_provider(provider)
-  end
-  if vim.fn.executable(provider._opts.command) ~= 1 then
-    Log:get_default().warn("Unable to find the path for", vim.inspect(provider))
-    return
+function M.registered_providers_name(ft)
+  local names = {}
+
+  for _, providers in pairs(M.langs[ft].registered) do
+    for _, provider in ipairs(providers) do
+      table.insert(names, provider.name)
+    end
   end
-  return provider._opts.command
+  return names
 end
 
 -- TODO: for linters and formatters with spaces and '-' replace with '_'
 function M.setup(filetype)
-  for _, formatter in pairs(lvim.lang[filetype].formatters) do
-    Log:get_default().debug("validating format provider: ", formatter.exe)
+  if not M.langs[filetype] then
+    M.langs[filetype] = {
+      registered = {
+        formatters = {},
+        linters = {},
+      },
+      errors = {
+        formatters = {},
+        linters = {},
+      },
+    }
+  end
```

Comment:
I think we should go a for a proper metatable because of this much attached info, at least they can provide some functions.
I had that idea originally but ended up scrapping it.